codec/src/main/java/io/netty/handler/codec/compression/Lz4FrameEncoder.java

```diff
@@ -163,6 +183,29 @@ private static int compressionLevel(int blockSize) {
     }
 
     @Override
+    protected ByteBuf allocateBuffer(ChannelHandlerContext ctx, ByteBuf msg, boolean preferDirect) {
+        int bufSize = 0;
+        int remaining = msg.readableBytes();
+        while (remaining > 0) {
+            int curSize = Math.min(blockSize, remaining);
+            remaining -= curSize;
+            // calculate the total compressed size of the current block (including header) and add to the total
+            bufSize += compressor.maxCompressedLength(curSize) + HEADER_LENGTH;
+        }
+
+        if (bufSize > maxEncodeSize) {
+            String errorMsg = "requested encode buffer size (%d bytes) exceeds the maximum allowable size (%d bytes)";
+            throw new EncoderException(String.format(errorMsg, bufSize, maxEncodeSize));
+        }
+
+        if (preferDirect) {
+            return ctx.alloc().ioBuffer(bufSize);
```

Comment:
we should also pass `maxEncodeSize` as the second parameter (e.g. `maxCapacity`) to these allocations. This will ensure the buffer will not grow beyond this max capacity.